File: Sorting and Searching/Array Division/ArrayDivision.cpp
#include<cstdio>
#include<algorithm>

using namespace std;

int n, k;
int a[200005];

bool check(long long x) {
    long long sum=0, cnt=0;
    for (int i=0; i<n;i++) {
        if (a[i]>x) return false;
        if (sum+a[i]>x) {
            sum=a[i];
            cnt++;
        }
        else sum += a[i];
    }
    return cnt < k;
}

int main() {
    scanf("%d %d", &n,&k);
    for (int i=0; i<n;i++) scanf("%d", &a[i]);
    long long l=1, r=1e18;
    while (l<r) {
        long long mid = l+(r-l)/2;
        if(check(mid)) r=mid;
        else l=mid+1;
    }
    printf("%lld\n", l);
    return 0;
}

File: Sorting and Searching/Array Division/ArrayDivision.py
import sys
# from time import time
# s = time()
input = sys.stdin.readline

def check(mid, k, arr):
    sum,cnt = 0,0
    for x in arr:
        if x>mid or cnt>=k:
            return False
        if sum+x>mid:
            cnt+=1
            sum = x
        else:
            sum+=x
    # if sum>0:
    #     cnt+=1
    return cnt < k

n,k = map(int, input().split())
arr = list(map(int, input().split()))
l,r = max(arr),sum(arr)

while l<r:
    mid = l+(r-l)//2
    if check(mid, k, arr):
        r = mid
    else:
        l = mid+1
        
sys.stdout.write(str(l)+'\n')
# sys.stderr.write("Time taken: "+str(time()-s)+'s\n')
